[He Outong] iP

IrvingHe000 opened the pull request:
# This is an update message using GFMD

> Here is a quote from somewhere.
> - somebody

## *Duke* is a very efficient tool that can help you store:
* Task
* Deadline
* Event
* Todo

## All you need to do is this:
1. Download from https://github.com/IrvingHe000/ip
2. Use IDEA to open it
3. Start to learn how to use it :smile:

And it is **FREE**

## Features
- [ ] add task
- [ ] delete task
- [ ] update task
- [ ] show taskList
- [ ] And so on

## Here is the main method
    public static void main(String[] args) {
            Application.launch(Duke.class, args);
    }

Commit:
Add Gradle support

IrvingHe000 committed:
Level-1

IrvingHe000 committed:
Level-2

IrvingHe000 committed:
Adding Task class, adding mark-as-done functionality

IrvingHe000 committed:
Adding subclasses of Task; change to toString() method

IrvingHe000 committed:
Check type of input (todo, event, deadline)

IrvingHe000 committed:
Adding testing by using shell script

IrvingHe000 committed:
Adding minimal exception handler

IrvingHe000 committed:
Change the storation of tasks from array to a collection

IrvingHe000 committed:
Adding delete task

IrvingHe000 committed:
Adding enum

IrvingHe000 committed:
Add methods of different operation on task

IrvingHe000 committed:
Add methods that can read tasks from and save tasks into a file

IrvingHe000 committed:
Change the storing of date and time of deadlineTask to Date and Time object

IrvingHe000 committed:
Merge branch 'branch-Level-7'

IrvingHe000 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java
#	src/main/java/Task.java

IrvingHe000 committed:
Make Duke more OOP by separating its methods into different classes

IrvingHe000 committed:
Divide classes into packages

IrvingHe000 committed:
add parser functionality to Parser class

IrvingHe000 committed:
Add JUnit test on Parser and TaskList

IrvingHe000 committed:
Fix a bug that sometimes duke.txt cannot be created

IrvingHe000 committed:
Add Javadoc to some public classes and methods.

IrvingHe000 committed:
Tweak the code to comply with a coding standard.

IrvingHe000 committed:
Add finding matching tasks method

IrvingHe000 committed:
Merge branch 'branch-A-JavaDoc'

IrvingHe000 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/command/InputTypes.java
#	src/main/java/duke/command/Parser.java
#	src/main/java/duke/command/Ui.java

IrvingHe000 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/command/InputTypes.java
#	src/main/java/duke/command/Ui.java

IrvingHe000 committed:
Add Command class

IrvingHe000 committed:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d' into A-Gradle

IrvingHe000 committed:
Add Gradle run and test

IrvingHe000 committed:
Merge branch 'A-Gradle'

IrvingHe000 committed:
Add basic GUI set-up according to the tutorial

fans2619 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,81 @@
+        this.storage = storage;
+    }
+
+    public void addNewTodo(String task) throws Exception {
```
Could it be better if you change the method name to `addNewTodoTask()`?

fans2619 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,81 @@
+        String target = task.substring(5);
+        List<Integer> result = new ArrayList<>();
+        for (int i = 0; i < taskList.size(); i++) {
+            Task cur = taskList.get(i);
```
Instead of using `cur` for simplicity, consider using `currentTask`?

fans2619 commented on src/main/java/duke/command/InputTypes.java:
```diff
@@ -0,0 +1,19 @@
+package duke.command;
+
+public enum InputTypes {
```
You may want to name it as `InputType` instead of using the plural form.

fans2619 commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    /**
+     * Test whether it has a valid Todo task format. If yes, return the
```
Coding standard: "Test**s** ..."

fans2619 commented on src/main/java/duke/command/Storage.java:
```diff
@@ -0,0 +1,108 @@
+            char taskType = task.charAt(0);
+            char taskDone = task.charAt(4);
+            String taskValue = task.substring(8);
+            boolean finished = (taskDone == '1');
```
Consider using something like `isFinished` for the `boolean` variable's name?

fans2619 commented on src/main/java/duke/command/TaskList.java:
```diff
@@ -0,0 +1,50 @@
+     * @param i The number of item that user wants to get.
+     * @return The task that user wants to get.
+     */
+    public Task get(int i) {
```
You may want to specify get WHAT for the method name in this case, e.g. `getTask`?

fans2619 commented on src/main/java/duke/command/TaskList.java:
```diff
@@ -0,0 +1,50 @@
+     * Add new task to the task list.
+     * @param task The new task user wants to add to the list.
+     */
+    public void add(Task task) {
```
Could `addTask()` be better? Similar for `removeTask()`?

fans2619 commented on src/main/java/duke/command/Ui.java:
```diff
@@ -0,0 +1,112 @@
+ */
+public class Ui {
+    private TaskList taskList;
+    private static String logo = " ____        _        \n"
```
I would recommend
```suggestion
    private static final String LOGO = " ____        _        \n"
```
although this line is given.

fans2619 commented on src/main/java/duke/command/Ui.java:
```diff
@@ -0,0 +1,112 @@
+        printSeparateLine();
+    }
+
+    public void showFindingTasks(List<Integer> nums) {
```
Could be improved.

fans2619 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,41 @@
+
+public class Task {
+    private String taskName;
+    private boolean done;
```
`isDone`

fans2619 reviewed (commented):
Well, those comments are just my own suggestions and they may not be good! 🙂

Nature711 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,85 @@
+                    command.findTask(task);
+                    break;
+                case UNKNOWN:
+                    throw new Exception("Cannot Understand");
```
Perhaps it's better to have customized classes for different kinds of exceptions, thereby customizing the exception message for each class. Then in the lines below you won't need to check if the exception message is equal to some message and then print out something else; instead you just need to throw an instance of that specific class of customized exception and the corresponding message can also be displayed.

Nature711 commented on src/main/java/duke/command/Storage.java:
```diff
@@ -0,0 +1,108 @@
+ * Dealing with loading tasks from the file and saving tasks in the file.
+ */
+public class Storage {
+    private File src;
```
Maybe it's better to declare this File src attribute as final?

Nature711 commented on src/main/java/duke/command/Storage.java:
```diff
@@ -0,0 +1,108 @@
+            fileWriter.write(task.toStoredString() + "\n");
+            fileWriter.close();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Instead of calling e.printStackTrace(), maybe it's better to create your own customized exception objects which will display more detailed and understandable information about the exception occured.

Nature711 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,41 @@
+        return done;
+    }
+
+    public void finished() {
```
Is it better to name this method to "isFinished"?

Nature711 reviewed (commented):
LGTM overall. The way in which exceptions are handled could be improved.

IrvingHe000 committed:
Modify GUI to act like Duke

IrvingHe000 committed:
Merge branch 'Level-10'

IrvingHe000 committed:
Add more java doc.

IrvingHe000 committed:
Add customised exceptions

juliussneezer04 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,49 @@
+
+    /**
+     * Adds new task to the task list.
+     * @param task The new task user wants to add to the list.
```
Maybe you could leave a new line between descriptions and parameters like in the styling guide!

juliussneezer04 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,145 @@
+    /**
+     * Get the GoodBye message.
+     *
+     * @return A StringBuilder containing the GoodBye message
```
Remember to add punctuation (like full-stops) in header descriptions!

juliussneezer04 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,71 @@
+     * @param done Whether the task is done or not.
+     * @param taskTime The time of Deadline.
+     */
+    public Deadline(String task, boolean done, String taskTime) {
```
Maybe you could consider changing boolean name to isDone?

juliussneezer04 commented on src/main/java/duke/command/Storage.java:
```diff
@@ -0,0 +1,128 @@
+                separatorPosition = taskValue.indexOf('|');
+                String deadlineName = taskValue.substring(0, separatorPosition - 1);
+                String deadlineTime = taskValue.substring(separatorPosition + 2);
+                curTask = new Deadline(deadlineName, finished, deadlineTime);
```
Don't forget to add 'break' statement at the end of cases! (or at least an explicit comment explaining its omission)

juliussneezer04 commented on src/main/java/duke/command/Storage.java:
```diff
@@ -0,0 +1,128 @@
+
+    /**
+     * Adds new task to the file when it is added to the taskList.
+     * @param task
```
Maybe elaborate on what the parameter is?

juliussneezer04 reviewed (commented):
Hi! I was tagged to review your code. It's mostly super clean 😄 with a few small errors, which I found with great difficulty! Keep it up man :)

IrvingHe000 committed:
Add assertions to the program

There are some places that a certain value or range of value are required.
Using assertions can help to warn users when their is a bug that those values
are not like what we expected.

This change improves the accuracy of the program because we can find bugs
more quickly.

Let's update the code by adding assertions when necessary.

IrvingHe000 committed:
Improve code quality

The code quality of this project was not so good in some aspects:
* Comments
* Annotation
* Naming

Let's improve the code quality according to the requirement.

Refer to this guide on how to improve code quality:
https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#code-quality

IrvingHe000 committed:
Merge pull request #1 from IrvingHe000/A-Assertions

Add assertions to the program

IrvingHe000 committed:
Merge branch 'master' into A-CodeQuality

IrvingHe000 committed:
Merge pull request #2 from IrvingHe000/A-CodeQuality

Improve code quality

IrvingHe000 committed:
Add Duke the ability to recognise and deal with duplicate items

Duke could not recognise duplicate items before, therefore, sometimes there will be
some same task appearing in the taskList.

Now if user adds duplicate item into the taskList, Duke will not add the new item into
the taskList. Instead, Duke will show a warning to tell the user that the task has already
existed.

IrvingHe000 committed:
Add duplicate exception

This exception is for handling duplicate elements.

Duke will show the message if user enters the same task into the task List.

IrvingHe000 committed:
Add more description of methods

IrvingHe000 committed:
Fix a bug that Todo task cannot be stored

IrvingHe000 committed:
Add date and time to Event

IrvingHe000 committed:
Add CML version of Duke

IrvingHe000 committed:
Change a bug may need todo to have empty name

IrvingHe000 committed:
Add exit program functionality for the GUI

IrvingHe000 committed:
Update Readme

IrvingHe000 committed:
Update method header

IrvingHe000 committed:
Update README.md to solve format problem
